File: services/adjudicate/attesta_adjudicate/manifest.py
"""InvestigationManifest (docs/ARCHITECTURE.md 2.9) and the subprocess
bridges to ledger/cmd/manifest_cli and ledger/cmd/merkle_cli -- the
Go-side counterparts of adjudicate_bridge.py's bridge to the Rust kernel.
Canonical CBOR encoding, BLAKE3 content addressing, and Ed25519 signing
all live in ledger/ (Go), same as every other evidence-path primitive in
this project; this module is the only place that has to know the two
CLIs' JSON shapes, exactly the discipline adjudicate_bridge.py already
established for the kernel.
"""

from __future__ import annotations

import json
import subprocess
from pathlib import Path
from typing import Any, Optional

from pydantic import BaseModel

ZERO_HASH = ""  # the documented "unset" sentinel manifest_cli's decodeHex32 accepts


class InferenceMeta(BaseModel):
    provider: str
    model_id: str
    weights_digest: str
    seed: Optional[int] = None
    decode_params: dict[str, str] = {}
    prompt_template_hashes: list[str] = []
    tool_versions: dict[str, str] = {}


class BudgetConsumed(BaseModel):
    tool_calls: int
    turns: int


class Manifest(BaseModel):
    investigation_id: str
    tenant_id: str
    case_id: str
    epoch_root: str = ZERO_HASH
    epoch_id: int = 0
    index_snapshot_hash: str = ZERO_HASH
    claim_set_hash: str
    claim_ids: list[str]
    inference: InferenceMeta
    policy_version: str
    kernel_version: str
    attack_model_version: str = "attack.unversioned"
    verdict_severity: str
    verdict_confidence: int
    verdict_disposition: str
    verdict_hash: str
    budget_consumed: BudgetConsumed
    started_at_ns: int
    completed_at_ns: int
    prev_manifest_hash: str = ZERO_HASH


class SealedManifest(BaseModel):
    manifest: Manifest
    manifest_hash: str
    signature: str
    public_key: str


class LedgerBridgeError(RuntimeError):
    pass


def find_go_cli(repo_root: Path, name: str) -> Path:
    names = [name, f"{name}.exe"]
    for profile in ("release", "debug"):
        for candidate_name in names:
            candidate = repo_root / "target" / profile / candidate_name
            if candidate.exists():
                return candidate
    raise FileNotFoundError(
        f"{name} binary not found under target/{{release,debug}}/ -- "
        f"run `go build -o target/release/{name} ./ledger/cmd/{name}` first"
    )


def _run_cli(cli_path: Path, request: dict[str, Any]) -> dict[str, Any]:
    proc = subprocess.run(
        [str(cli_path)],
        input=json.dumps(request),
        capture_output=True,
        text=True,
        timeout=30,
    )
    try:
        output = json.loads(proc.stdout)
    except json.JSONDecodeError as e:
        raise LedgerBridgeError(
            f"{cli_path.name} produced non-JSON output (exit {proc.returncode}): "
            f"stdout={proc.stdout!r} stderr={proc.stderr!r}"
        ) from e
    if proc.returncode != 0:
        raise LedgerBridgeError(f"{cli_path.name} rejected the request: {output.get('error', proc.stderr)}")
    return output


def seal_manifest(manifest: Manifest, signing_key_seed_hex: str, manifest_cli_path: Path) -> SealedManifest:
    """Canonicalizes, hashes, and signs `manifest` via manifest_cli. The
    caller is responsible for setting manifest.prev_manifest_hash to the
    tenant's current chain tip before calling this -- see store.py's
    ManifestStore, which owns that bookkeeping so this function can stay a
    thin, stateless bridge, matching ledger.SealManifest's own purity on
    the Go side.
    """
    result = _run_cli(
        manifest_cli_path,
        {"op": "seal", "signing_key_seed": signing_key_seed_hex, "manifest": manifest.model_dump(mode="json")},
    )
    return SealedManifest(
        manifest=manifest,
        manifest_hash=result["manifest_hash"],
        signature=result["signature"],
        public_key=result["public_key"],
    )


def verify_sealed_manifest(sealed: SealedManifest, manifest_cli_path: Path) -> bool:
    """Re-derives the manifest's hash from its own fields and checks both
    the hash and the Ed25519 signature -- the exact check a third party
    runs against an exported bundle with no access to any of ATTESTA's own
    systems. Never trusts `sealed.manifest_hash` at face value: it is
    recomputed from `sealed.manifest` inside manifest_cli itself.
    """
    result = _run_cli(
        manifest_cli_path,
        {
            "op": "verify",
            "manifest": sealed.manifest.model_dump(mode="json"),
            "manifest_hash": sealed.manifest_hash,
            "signature": sealed.signature,
            "public_key": sealed.public_key,
        },
    )
    return bool(result["valid"])


def canon_hash(value: Any, manifest_cli_path: Path) -> str:
    """Canonicalizes and BLAKE3-hashes an arbitrary JSON-serializable
    value via manifest_cli's "canon_hash" op -- used for
    Manifest.claim_set_hash, which commits to the submitted claim set as a
    single content address independent of the kernel's own per-claim
    claim_id() (see kernel/src/bin/adjudicate_cli.rs's
    `submitted_claim_ids`, a separate, complementary identity).
    """
    result = _run_cli(manifest_cli_path, {"op": "canon_hash", "value": value})
    return str(result["hash"])


class MerkleProofStep(BaseModel):
    hash: str
    is_left: bool


class InclusionProof(BaseModel):
    leaf_index: int
    steps: list[MerkleProofStep]
    root: str


def merkle_prove(evidence_ids: list[str], leaf_index: int, merkle_cli_path: Path) -> InclusionProof:
    result = _run_cli(merkle_cli_path, {"op": "prove", "evidence_ids": evidence_ids, "leaf_index": leaf_index})
    return InclusionProof(
        leaf_index=result["leaf_index"],
        steps=[MerkleProofStep(**s) for s in result["steps"]],
        root=result["root"],
    )


def merkle_verify(evidence_id: str, proof: InclusionProof, root: str, merkle_cli_path: Path) -> bool:
    result = _run_cli(
        merkle_cli_path,
        {
            "op": "verify",
            "evidence_id": evidence_id,
            "leaf_index": proof.leaf_index,
            "steps": [s.model_dump() for s in proof.steps],
            "root": root,
        },
    )
    return bool(result["valid"])

File: ledger/manifest.go
package ledger

import (
	"crypto/ed25519"

	"lukechampine.com/blake3"
)

// InferenceMeta records which model/provider produced a claim set —
// docs/ARCHITECTURE.md §2.9's `inference` block. DecodeParams is
// string-keyed/string-valued rather than numeric: decode parameters like
// temperature are inherently fractional, and this package's canonical
// encoding deliberately has no float rule (see cbor.go's doc comment) —
// keeping them as their literal string form sidesteps re-opening that
// question for a field that is descriptive metadata, not something the
// verdict depends on.
type InferenceMeta struct {
	Provider             string
	ModelID              string
	WeightsDigest        string
	Seed                 *int64
	DecodeParams         map[string]string
	PromptTemplateHashes [][32]byte
	ToolVersions         map[string]string
}

// BudgetConsumed is docs/ARCHITECTURE.md §2.9's `budget_consumed` — how
// much of services/investigate's Budget (Phase 6) an investigation used.
type BudgetConsumed struct {
	ToolCalls int64
	Turns     int64
}

// Manifest is docs/ARCHITECTURE.md §2.9's InvestigationManifest, minus the
// two fields (`manifest_hash`, `signature`) that only exist once the
// manifest has been sealed — see SealedManifest.
//
// IndexSnapshotHash is carried as an opaque field and left the zero value
// in this phase: a real full-text/vector index over case state doesn't
// exist yet (no phase before Phase 9's UI needs it), so there is nothing
// honest to hash into it yet. It is part of the struct now so sealing a
// manifest never has to change shape again once the index exists.
type Manifest struct {
	InvestigationID    string
	TenantID           string
	CaseID             string
	EpochRoot          [32]byte
	EpochID            uint64
	IndexSnapshotHash  [32]byte
	ClaimSetHash       [32]byte
	ClaimIDs           [][32]byte
	Inference          InferenceMeta
	PolicyVersion      string
	KernelVersion      string
	AttackModelVersion string
	VerdictSeverity    string
	VerdictConfidence  int64
	VerdictDisposition string
	VerdictHash        [32]byte
	BudgetConsumed     BudgetConsumed
	StartedAtNS        int64
	CompletedAtNS      int64
	// PrevManifestHash chains this manifest to the previous one sealed for
	// the same TenantID — docs/ARCHITECTURE.md §2.9: "hash-chained per
	// tenant." The zero value marks the first manifest in a tenant's
	// chain. Callers own tracking "what was the last hash for this
	// tenant" (see services/adjudicate/attesta_adjudicate/store.py) —
	// this package stays a pure function of its input, the same
	// discipline epoch.go's SealEpoch already applies to its own `prev`
	// parameter.
	PrevManifestHash [32]byte
}

// SealedManifest is a Manifest plus the two fields that exist only once
// it's been canonicalized, hashed, and signed.
type SealedManifest struct {
	Manifest
	ManifestHash [32]byte
	Signature    []byte
	PublicKey    ed25519.PublicKey
}

// canonicalManifestMap builds the map[string]any that EncodeCanonical
// hashes. Field order in this function is irrelevant to the resulting
// bytes — encodeMap sorts by encoded key regardless — but every field is
// listed explicitly and by name specifically so adding a struct field
// without adding it here is a visible diff, not a silent gap in what
// actually gets committed to the hash chain.
func canonicalManifestMap(m Manifest) map[string]any {
	claimIDs := make([]any, len(m.ClaimIDs))
	for i, id := range m.ClaimIDs {
		claimIDs[i] = id[:]
	}
	promptHashes := make([]any, len(m.Inference.PromptTemplateHashes))
	for i, h := range m.Inference.PromptTemplateHashes {
		promptHashes[i] = h[:]
	}
	decodeParams := make(map[string]any, len(m.Inference.DecodeParams))
	for k, v := range m.Inference.DecodeParams {
		decodeParams[k] = v
	}
	toolVersions := make(map[string]any, len(m.Inference.ToolVersions))
	for k, v := range m.Inference.ToolVersions {
		toolVersions[k] = v
	}
	var seed any
	if m.Inference.Seed != nil {
		seed = *m.Inference.Seed
	}

	return map[string]any{
		"investigation_id":    m.InvestigationID,
		"tenant_id":           m.TenantID,
		"case_id":             m.CaseID,
		"epoch_root":          m.EpochRoot[:],
		"epoch_id":            uint64(m.EpochID),
		"index_snapshot_hash": m.IndexSnapshotHash[:],
		"claim_set_hash":      m.ClaimSetHash[:],
		"claim_ids":           claimIDs,
		"inference": map[string]any{
			"provider":               m.Inference.Provider,
			"model_id":               m.Inference.ModelID,
			"weights_digest":         m.Inference.WeightsDigest,
			"seed":                   seed,
			"decode_params":          decodeParams,
			"prompt_template_hashes": promptHashes,
			"tool_versions":          toolVersions,
		},
		"policy_version":       m.PolicyVersion,
		"kernel_version":       m.KernelVersion,
		"attack_model_version": m.AttackModelVersion,
		"verdict_severity":     m.VerdictSeverity,
		"verdict_confidence":   m.VerdictConfidence,
		"verdict_disposition":  m.VerdictDisposition,
		"verdict_hash":         m.VerdictHash[:],
		"budget_consumed": map[string]any{
			"tool_calls": m.BudgetConsumed.ToolCalls,
			"turns":      m.BudgetConsumed.Turns,
		},
		"started_at_ns":      m.StartedAtNS,
		"completed_at_ns":    m.CompletedAtNS,
		"prev_manifest_hash": m.PrevManifestHash[:],
	}
}

// ManifestHash computes BLAKE3(canonical_cbor(manifest)) — the content
// address a SealedManifest commits to, and the value the next manifest in
// the same tenant's chain sets as its own PrevManifestHash.
func ManifestHash(m Manifest) ([32]byte, error) {
	canonical, err := EncodeCanonical(canonicalManifestMap(m))
	if err != nil {
		return [32]byte{}, err
	}
	return blake3.Sum256(canonical), nil
}

// SealManifest computes the manifest's hash and signs it with the given
// per-tenant Ed25519 key — the same signing primitive epoch.go's
// SealEpoch already uses, applied here to a manifest instead of an epoch
// header. A pure function: it never consults or mutates any store, so the
// caller supplies PrevManifestHash on m already set correctly (see the
// Manifest.PrevManifestHash doc comment).
func SealManifest(m Manifest, signer ed25519.PrivateKey) (SealedManifest, error) {
	hash, err := ManifestHash(m)
	if err != nil {
		return SealedManifest{}, err
	}
	sig := ed25519.Sign(signer, hash[:])
	return SealedManifest{
		Manifest:     m,
		ManifestHash: hash,
		Signature:    sig,
		PublicKey:    signer.Public().(ed25519.PublicKey),
	}, nil
}

// VerifySealedManifest recomputes the manifest's hash from its own fields
// and checks both that it matches the claimed ManifestHash (nothing was
// tampered with after sealing) and that the signature verifies against
// the embedded public key. This is the exact check a third party runs
// against an exported bundle with no access to any of ATTESTA's own
// systems — docs/ARCHITECTURE.md §2.9's "court-defensible" property.
func VerifySealedManifest(sm SealedManifest) (bool, error) {
	recomputed, err := ManifestHash(sm.Manifest)
	if err != nil {
		return false, err
	}
	if recomputed != sm.ManifestHash {
		return false, nil
	}
	return ed25519.Verify(sm.PublicKey, sm.ManifestHash[:], sm.Signature), nil
}
